File: problems/problems_777/solution.py
import solution
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.canTransform(*test_input)

    def canTransform(self, start: str, result: str) -> bool:
        i, j = 0, 0
        n = len(start)
        while i < n or j < n:
            while i < n and start[i] == 'X':
                i += 1
            while j < n and result[j] == 'X':
                j += 1
            if i == n or j == n:
                return i == j
            if start[i] != result[j]:
                return False
            if start[i] == 'L' and i < j:
                return False
            if start[i] == 'R' and i > j:
                return False
            i += 1
            j += 1
        return i == j

File: problems/problems_777/Solution.java
package problems.problems_777;

import com.alibaba.fastjson.JSON;
import java.util.*;
import qubhjava.BaseSolution;


public class Solution extends BaseSolution {
    public boolean canTransform(String start, String result) {
        int l = 0, r = 0;
        int n = start.length();
        for (int i = 0; i < n; ++i) {
            if ((start.charAt(i) == 'L' && r != 0) || (start.charAt(i) == 'R' && l != 0)) {
                return false;
            }
            l += start.charAt(i) == 'L' ? 1 : 0;
            r += start.charAt(i) == 'R' ? 1 : 0;
            l -= result.charAt(i) == 'L' ? 1 : 0;
            r -= result.charAt(i) == 'R' ? 1 : 0;
            if ((l != 0 && r != 0) || l > 0 || r < 0) {
                return false;
            }
        }
        return l == 0 && r == 0;
    }

    @Override
    public Object solve(String[] inputJsonValues) {
        String start = jsonStringToString(inputJsonValues[0]);
		String result = jsonStringToString(inputJsonValues[1]);
        return JSON.toJSON(canTransform(start, result));
    }
}
